### ISSUE TITLE
tf.keras.layers.LSTM not converted to ONNX LSTM layer

### ISSUE THREAD

**Issue by swartjean** [onnx/tensorflow-onnx#1546]
**Describe the bug**
I am working with a Keras model that contains LSTM layers and I have been unable to use tf2onnx to convert it to an ONNX model with matching LSTM layers. Minimal example below:

```python
import tensorflow as tf
import tf2onnx

# Construct model with a single LSTM layer
model = tf.keras.models.Sequential()
model.add(
    tf.keras.layers.LSTM(
        64,
        return_sequences=True,
        input_shape=(128, 1),
    )
)
model.compile()

# Save the Keras model
model.save("tf2onnx_keras_model.h5")

# Convert to ONNX and save
onnx_model, external_tensor_storage = tf2onnx.convert.from_keras(
    model, output_path="tf2onnx_onnx_model.onnx"
)
```

My system configuration:

OS: Windows 10
Tensorflow Version: 2.4.1
tf2onnx Version: 1.8.5
Python Version: 3.8.10

I have also tried specifying `opset=13` with no luck.

Viewing the resultant Keras and ONNX models in Netron reveals:

![tf2onnx_netron](https://user-images.githubusercontent.com/16536996/120456709-277dd200-c396-11eb-8efd-a394fb8395d6.png)

I have also tested [keras2onnx](https://github.com/onnx/keras-onnx) using Tensorflow 2.2.2 (keras2onnx does not appear to have TF 2.4.x compatibility yet). Minimal example below:

```python
import tensorflow as tf
import keras2onnx

# Construct model with a single LSTM layer
model = tf.keras.models.Sequential()
model.add(
    tf.keras.layers.LSTM(
        64,
        return_sequences=True,
        input_shape=(128, 1),
    )
)
model.compile()

# Save the Keras model
model.save("keras2onnx_keras_model.h5")

# Convert to ONNX and save
onnx_model = keras2onnx.convert_keras(model)
keras2onnx.save_model(onnx_model, "keras2onnx_onnx_model.onnx")
```

My system configuration for this example:

OS: Windows 10
Tensorflow Version: 2.2.2
keras2onnx Version: 1.7.0
Python Version: 3.8.10

Viewing the resultant Keras and ONNX models in Netron shows that the Keras LSTM layer was converted into an ONNX LSTM layer:

![keras2onnx_netron](https://user-images.githubusercontent.com/16536996/120456704-264ca500-c396-11eb-8c7a-f23f6b839d01.png)

Is there something that I am missing in the model specification or the conversion process that is needed for tf2onnx to properly convert LSTM nodes?

At the moment the converted tf2onnx model exhibits worse inference speed than the Keras model, especially when running with the CUDA execution provider. I suspect this is due to shape and slice operations falling back to the CPU, causing a lot of unnecessary device-host and host-device memory transfers. I have not tested thoroughly, but it does not look like the ONNX model generated with keras2onnx exhibits the same problems - but reverting back to TF 2.2.2 is not a great solution.

Thanks in advance!

**Comment by guschmue** [comment]
for tf-1.x we try this half baked but for tf-2.x we use the v2 control flow and are not even trying. The rational was that we rarely see lstm's simple enough for this to work. As soon they come with attention and other features the onnx lstm would not work anyway.
But we can take a look.

**Comment by swartjean** [comment]
@guschmue thanks for clarifying. Like I mentioned - I have seen a significant difference in inference speed between the two conversions, so if this could remain as a low-priority feature request I would very much appreciate it.

**Comment by q-ycong-p** [comment]
Hi @TomWildenhain-Microsoft and @guschmue,

Thanks for working on this issue. However I still encounter issue reported in the original post:
1. I've pulled the latest master and confirmed that it containes [#1584](https://github.com/onnx/tensorflow-onnx/pull/1584). I saved the keras model exactly as in @swartjean's example, and ran `python -m tf2onnx.convert --keras tf2onnx_keras_model.h5  --output tf2onnx_out.onnx`. The resultant onnx model still contains loop instead of LSTM,  as shown below. 
I've tried appending `--opset 13` to the command, and tried running `export TF_KERAS=1` but still get the same onnx model that contains loop. Could you look at this issue please?


![Screen Shot 2021-09-20 at 4 34 27 PM](https://user-images.githubusercontent.com/50014064/134090077-f81781d1-8716-46cb-9d08-beec1aeee4b5.png)




2. In the above-mentioned case, I've also saved the intermediate frozen graph using the `--output_frozen_graph frozen.pb`, which looks like:

![Screen Shot 2021-09-20 at 5 09 00 PM](https://user-images.githubusercontent.com/50014064/134092939-5349eed6-3458-4126-8d90-92b2904f5bd9.png)

Could you confirm for me that it is the correct way to save a frozen graph? The reason for asking, is that for more completed model that contains LSTM, the frozen graph looks difficult to interpret and almost seems wrong. 

Thank you!

**Comment by buddhapuneeth** [comment]
@TomWildenhain-Microsoft I am also seeing this issue on TF 2.5 and TF 2.3. 
I would like to know is there any specific versions of dependencies we need to try?

**Comment by TomWildenhain-Microsoft** [comment]
Hi @buddhapuneeth and @swartjean 

Unfortunately I think the conversion I added only works for the LSTMCell pattern.  We will need to add a pattern for the keras LSTM layer (likely a few for the different cases).  I'm a bit busy with other tasks currently, but if you want to help, what we need is something similar to this for LSTMs: https://github.com/onnx/tensorflow-onnx/pull/1692/files

**Comment by q-ycong-p** [comment]
@TomWildenhain-Microsoft Thank you. I'll use the keras GRU changes as guide to add pattern for keras LSTM layer. Will contribute back once works.

**Comment by TomWildenhain-Microsoft** [comment]
@q-ycong-p Thanks so much.  It will be a bit tricky.  The currently code is a bit messy since it tries to reuse the same patterns for tf1 and tf2.  tf1 has inline loops while tf2 nests them but the cores are the same.  It's fine to implement these new patterns for just tf2 for now.  You will want to add a new pattern to RNN utils then reference it in the tf2 lstm rewriter.  Finally add a unit test and the CI will check it against all tf2 versions (make sure to add a min tf version decorator).

**Comment by q-ycong-p** [comment]
@TomWildenhain-Microsoft ,

Thank you for the direction - appreciate it! Could you pls help me understand this to proceed:
The saved frozen graphs (.pb) parsed from both tf.compat.v1.nn.rnn_cell.LSTMCell and tf.keras.layers.LSTM models, does not visualize useful node information, such as that of Mul, tanh and sigmoid in Netron. It looks like below. No further node info within the while loop is made available. In this case, what's the best way to understand what the keras LSTM we want to match to?

I understand that after internal parsing, the subgraph contains the above-said node types. However, using debugger to run through these while loops is not a realistic way to understand the generic LSTM pattern. Any suggestion? 

![Screen Shot 2021-09-23 at 10 07 14 AM](https://user-images.githubusercontent.com/50014064/134552448-fc1fde6b-5630-472f-b755-ff758180414f.png)

**Comment by q-ycong-p** [comment]
Specifically, any way we can save a subgraph of type [tf2onnx.graph.Graph](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/graph.py#L442) and visualize it somehow?

**Comment by TomWildenhain-Microsoft** [comment]
Click on the while loop, then click the subgraph name.  It should open the subgraph.

**Comment by q-ycong-p** [comment]
> Click on the while loop, then click the subgraph name. It should open the subgraph.

Thanks for reply. The above screenshot is already the view after clicking 'while' loop in graph. As we see, the right-hand-side drop-down does not contain option to open subgraph. 

Please note that this is the [frozen graph of tf_graph](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/convert.py#L159)[1], before it is processed internally to [tf2onnx.graph.Graph ](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/graph.py#L442). Thus it might not already has nodes of useful types. 

Since we use [tf2onnx.graph.Graph ](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/graph.py#L442) for graph manipulation and pattern-matching, it would help to directly visualize the nodes in it. So far, I manage to dump node info with [this](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/graph.py#L1236)[2]. But it's difficult to untangle generic pattern of keras LSTM in such way. 

What have people been doing to view custom patterns they want to match?

I've attached the [1] above-said frozen graph and the [2] prints from [Graph.dump_graph()](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/graph.py#L1236) below for your reference:

[test_single_dynamic_lstm_time_major.pb.zip](https://github.com/onnx/tensorflow-onnx/files/7220674/test_single_dynamic_lstm_time_major.pb.zip)


![Screen Shot 2021-09-23 at 11 35 03 AM](https://user-images.githubusercontent.com/50014064/134579131-794d5688-0b9d-4cd0-8f1e-a8612e1c3de7.png)

**Comment by TomWildenhain-Microsoft** [comment]
I'm on Netron version 5.2.1 and I can click directly on the subgraph name of the file you attached to view it.

![image](https://user-images.githubusercontent.com/67606533/134580998-8f9f0e1c-3522-404e-a508-b9faea96d284.png)

![image](https://user-images.githubusercontent.com/67606533/134581048-dde82f9d-4cf1-42ed-8cec-5b747f397f99.png)

**Comment by q-ycong-p** [comment]
> I'm on Netron version 5.2.1 and I can click directly on the subgraph name of the file you attached to view it.
> 
> ![image](https://user-images.githubusercontent.com/67606533/134580998-8f9f0e1c-3522-404e-a508-b9faea96d284.png)
> 
> ![image](https://user-images.githubusercontent.com/67606533/134581048-dde82f9d-4cf1-42ed-8cec-5b747f397f99.png)

omg this works perfectly. You are awesome. Thanks!

**Comment by q-ycong-p** [comment]
Hi @TomWildenhain-Microsoft ,

I made some progress following your suggestion on this implementation. I am able have keras LSTM patterns recognized and have the rewriter process tensors correctly (i.e. model yields correct results). 

However, I am seeing the init_h and init_c being passed to LSTM via an unexpected path, as show below (right). After some digging, it looks like this is become h and c are [separate](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/rewriter/lstm_tf2_rewriter.py#L65), which leads to [state_is_tuple=True](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/rewriter/lstm_tf2_rewriter.py#L72) route being used. This itself does not introduce incorrectness to inference results, but the converted onnx graph looks unnecessarily redundant compared to what keras2onnx produces, see below (left). 

I wonder what is the significance of have path (slice->concat->unsqeeze) passing (init-h, init_c) to LSTM. 
Can we assume that correct conversion should produce the exact same graph as keras2onnx does? Thank you!


![Screen Shot 2021-09-30 at 5 26 23 PM](https://user-images.githubusercontent.com/50014064/135548093-4e159e48-deb0-4b5d-a6d8-298aa5e012bf.png)

**Comment by TomWildenhain-Microsoft** [comment]
@q-ycong-p that looks great. Notice that the path you are referring to is all under a Shape op. Are you sure this is from the state is tuple path?  It looks to me like it is just assembling the Shape to use for Expand (create the initial state).  The discrepancy is likely that keras2onnx detects a default start state of 0, while tf2onnx (or tf) assembles a zero tensor (by expanding a 0 value to a shape computed from the input).

The Concat node from the state_is_tuple case comes after the LSTM node.
https://github.com/onnx/tensorflow-onnx/blob/446494eea2fb80b4beca13914811678399c91e11/tf2onnx/rewriter/lstm_tf2_rewriter.py#L189

The most important thing is you got rid of the loop node.  I'm very happy with what you have and expect it to be essentially as fast as the keras2onnx result.

**Comment by q-ycong-p** [comment]
@TomWildenhain-Microsoft Thank you for the detailed explanation! Now I understand the extra path on assembling shape. 

As I start writing test cases for this implementation, I notice:
* The keras2onnx_unit_tests imports keras via `tensorflow.python.keras` ([code](https://github.com/onnx/tensorflow-onnx/blob/master/tests/keras2onnx_unit_tests/mock_keras2onnx/proto/__init__.py#L41)) instead of the public `tensorflow.keras`. The former is known to be private and bug-prone, see thread in [Tensorflow](https://github.com/tensorflow/tensorflow/issues/33075#issuecomment-539070546) and [discussion](https://stackoverflow.com/questions/58279628/what-is-the-difference-between-tf-keras-and-tf-python-keras). Besides, official keras page(s) also refers to the later usage, see [here](https://keras.io/api/layers/). 
* I understand this discrepancy comes from keras2onnx, but I propose that we consider testing against `tensorflow.keras` moving forward?
* The reason why importing keras via `tensorflow.python.keras` blocks current development is:

For a Simple Keras LSTM layer:
1. Expected frozen graph that `tensorlfow.keras` generates shown below. Notice the sigmoid op shown correctly.
![Screen Shot 2021-10-01 at 10 31 19 AM](https://user-images.githubusercontent.com/50014064/135662809-14004875-e054-471a-9223-1cdf0bfb1fdf.png)

2. Unexpected frozen graph that `tensorflow.python.keras `generates shown below. Notice the sigmoid op replaced by strange (Mul->Add->Min->Max) sequence. I feel that this frozen graph makes less sense, and that we might not want/need to support it. 
![Screen Shot 2021-10-01 at 10 33 02 AM](https://user-images.githubusercontent.com/50014064/135663019-1853221a-56fb-4d53-ba21-7309928ad8a1.png)



I have also tested that, replacing tensorflow.python.keras by tensorflow.keras in keras2onnx_unit_tests does not break any existing unit tests, after fixing some legacy imports. For example, replace `keras.layers.advanced_activations` with individual `keras.layers`.

Do you think we can make this change on tf2onnx side?

**Comment by TomWildenhain-Microsoft** [comment]
Yes, I don't see a problem with updating it if you can. Submit a PR and see if it works. There are a bunch of old tf/keras versions the CI runs so we'll see if there are any bugs.

That said, the main function of those tests was to ensure parity with keras2onnx. New tests should go in the normal tf2onnx test locations (test_lstm.py in this case). These tests also will convert the model to tflite and tfjs and test the results, and try a lot more tf opsets.

See this one: https://github.com/onnx/tensorflow-onnx/blob/446494eea2fb80b4beca13914811678399c91e11/tests/test_gru.py#L720

The legacy keras2onnx tests are just for correctness. If you want to see whether you get rid of loops there too, add a no_loops_in_tf2 assertion to an existing keras2onnx lstm test. But if it doesn't work there for reasons like the issue you mention, that's fine too.

The specific problem you mention might be a different activation function being used by default. Support for all activation functions is not a priority since we should already convert everything correctly anyway, it will just be slower and use loops if the rewriter doesn't match.

**Comment by q-ycong-p** [comment]
@TomWildenhain-Microsoft Got it, thank you!

**Comment by TomWildenhain-Microsoft** [comment]
@q-ycong-p Can you open a PR with what you have so far?
